File: done/jvm/java/src/main/java/madness/DummyLogic.java
package madness;

import madness.entities.Name;
import madness.entities.Permission;
import madness.entities.Subscription;
import madness.entities.User;

import java.util.ArrayList;
import java.util.List;

public class DummyLogic {
    public static String getUserIdFromToken(String tocken) {
        return tocken.split("_")[1];
    }

    public static User createUser(String id, Name newName) {
        Name name = new Name("FirstName", "Surname");
        return new User(id, newName != null ? newName : name);
    }

    public static User.Status getUserStatus(String userId) {
        return User.Status.Active;
    }

    public static Subscription getUserSubscription(String userId) {
        return new Subscription(Subscription.Type.Basic);
    }

    public static List<Permission> getPermissions(User.Status status, Subscription.Type subscription) {
        List<Permission> permissions = new ArrayList<>(5);

        // default
        permissions.add(new Permission(Permission.Type.Cry));

        // active user permissions
        if (status == User.Status.Active) {
            permissions.add(new Permission(Permission.Type.ChangeName));

            // subscription based
            if (subscription != Subscription.Type.None) {
                permissions.add(new Permission(Permission.Type.WriteComments));
            }

            if (subscription == Subscription.Type.Premium) {
                permissions.add(new Permission(Permission.Type.CancelPremium));
                permissions.add(new Permission(Permission.Type.KickChildren));
            }
        }

        return permissions;
    }
}

File: node/src-kt/dummy_logic.kt
package madness

// Just imagine that we have different servises for users, statuses (antifrod) and subscriptions (billing)

fun getUserIdFromToken(tocken: String): String {
    return tocken.substringAfter("_")
}

fun createUser(id: String, newName: Name? = null): User {
    val name = Name("FirstName", "Surname")
    return User(id, newName ?: name)
}

fun getUserStatus(userId: String): User.Status {
    return User.Status.Active
}

fun getUserSubscription(userId: String): Subscription {
    return Subscription(Subscription.Type.Basic)
}

fun getPermissions(status: User.Status, subscription: Subscription.Type): List<Permission> {
    val permissions = mutableListOf(Permission(Permission.Type.Cry))

    // active user permissions
    if (status == User.Status.Active) {
        permissions.add(Permission(Permission.Type.ChangeName))

        // subscription based
        if (subscription != Subscription.Type.None) {
            permissions.add(Permission(Permission.Type.WriteComments))
        }

        if (subscription == Subscription.Type.Premium) {
            permissions.add(Permission(Permission.Type.CancelPremium))
            permissions.add(Permission(Permission.Type.KickChildren))
        }
    }

    return permissions
}